StripePaymentSheet/StripePaymentSheet/Source/Internal/Link/Utils/LinkURLGenerator.swift

```diff
@@ -4,36 +4,107 @@
 
 struct LinkURLParams: Encodable {
     struct MerchantInfo: Encodable {
-        let businessName: String
-        let country: String
+        var businessName: String
+        var country: String
     }
     struct CustomerInfo: Encodable {
-        let country: String
-        let email: String
+        var country: String
+        var email: String?
     }
     struct PaymentInfo: Encodable {
-        let currency: String
-        let amount: Int
-    }
-    let path = "mobile_pay"
-    let integrationType = "mobile"
-    let publishableKey: String
-    let merchantInfo: MerchantInfo
-    let customerInfo: CustomerInfo
-    let paymentInfo: PaymentInfo
-    let returnUrl: URL
-    let experiments: [String]
-    let flags: [String]
-    let loggerMetadata: [String]
-    let locale = Locale.current.toLanguageTag()
+        var currency: String
+        var amount: Int
+    }
+    enum PaymentObjectMode: String, Encodable {
+        case link_payment_method
+        case card_payment_method
+    }
+    var path = "mobile_pay"
+    var integrationType = "mobile"
+    var paymentObject: PaymentObjectMode
+    var publishableKey: String
+    var stripeAccount: String?
+    var paymentUserAgent: String
+    var merchantInfo: MerchantInfo
+    var customerInfo: CustomerInfo
+    var paymentInfo: PaymentInfo?
+    var experiments: [String: Bool]
+    var flags: [String: Bool]
+    var loggerMetadata: [String: Bool]
+    var locale: String
 }
 
 class LinkURLGenerator {
-    static func url() -> URL {
-//      TODO: Create URLs based on params
-        return URL(string: "https://checkout.link.com/#")!
+    static func linkParams(configuration: PaymentSheet.Configuration, intent: Intent) async throws -> LinkURLParams {
+        guard let publishableKey = configuration.apiClient.publishableKey ?? STPAPIClient.shared.publishableKey else {
+            throw LinkURLGeneratorError.noPublishableKey
+        }
+        guard let merchantCountryCode = intent.countryCode else {
+            throw LinkURLGeneratorError.noCountryCode
+        }
+
+        // US isn't the best default, but we only expect regionCode to be nil in the nil region on a simulator, and we don't want to crash.
+        let customerCountryCode = configuration.defaultBillingDetails.address.country ?? Locale.current.regionCode ?? "US"
+
+        // Get email from the billing details, or the Customer object if the billing details are empty
+        var customerEmail = configuration.defaultBillingDetails.email
+        if customerEmail == nil,
+           let customerID = configuration.customer?.id,
+           let ephemeralKey = configuration.customer?.ephemeralKeySecret,
+           let customer = try? await configuration.apiClient.retrieveCustomer(customerID, using: ephemeralKey)
+        {
+            customerEmail = customer.email
+        }
+
+        let merchantInfo = LinkURLParams.MerchantInfo(businessName: configuration.merchantDisplayName, country: merchantCountryCode)
```

Comment:
These args will all be base64 encoded, right? So we don't have to worry about non-encodable characters in our URL?

Reply:
yes